osquery/core/windows/bstr.cpp

```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/utils/system/system.h>
+
+#include <oleauto.h>
+
+#include <cassert>
+#include <osquery/core/windows/bstr.h>
+
+namespace osquery {
+
+namespace {
+static_assert(sizeof(Bstr) == sizeof(BSTR), "BstrSize");
+}
+
+// static
+Bstr Bstr::fromString(std::wstring_view s) {
+  return Bstr(SysAllocString(s.data()));
+}
+
+Bstr::~Bstr() {
+  // SysFreeString handles nullptr gracefully.
+  ::SysFreeString(bstr_);
+}
+
+Bstr::Bstr(BSTR bs) : bstr_(bs) {}
+
+Bstr::Bstr(Bstr&& other) : bstr_(other.bstr_) {
+  other.bstr_ = nullptr;
+}
+
+Bstr& Bstr::operator=(Bstr&& other) {
+  reset(other.release());
+  return *this;
+}
+
+void Bstr::reset(BSTR bstr) {
+  if (bstr != bstr_) {
+    ::SysFreeString(bstr_);
+    bstr_ = bstr;
+  }
+}
+
+BSTR Bstr::release() {
+  BSTR bstr = bstr_;
+  bstr_ = nullptr;
+  return bstr;
+}
+
+BSTR* Bstr::receiveAddress() {
+  assert(!bstr_);
```

Comment:
add a comment on assert usage. Also, I'm not a big fan of adding asserts on production code, I'd prefer to return gracefully if possible.

Reply:
Main reason to add this `assert` is to prevent memory leak and to ensure that the `Bstr::bstr_` pointer is null and does not hold any `BSTR` buffer allocated earlier.
Consider the following scenario:
```
Bstr bs = Bstr:: fromString(L"foo");
GetFileTime(..., bs.receiveAddress()); // for Debug builds assert can signal here
```

I personally like the idea of some kind of debug macro helpers with a well defined behaviour like [DCHECK](https://chromium.googlesource.com/chromium/src/+/main/base/check.h#23) and I wasn't sure what to use in this project.
Another approach is to free the buffer every time `receiveAddress` is called.

Reply:
good, thanks for the explanation! You can add this information to a code comment